x-pack/plugins/security_solution/server/lib/telemetry/async_sender.ts

```diff
@@ -0,0 +1,397 @@
+  }
+
+  public updateDefaultQueueConfig(config: QueueConfig): void {
+    this.fallbackQueueConfig = cloneDeep(config);
+  }
+
+  // internal methods
+  private queue$(
+    upstream$: rx.Observable<Event>,
+    channel: TelemetryChannel,
+    send: (channel: TelemetryChannel, events: string[]) => Promise<Result>
+  ): rx.Observable<Result> {
+    let inflightEventsCounter: number = 0;
+    const inflightEvents$: rx.Subject<number> = new rx.Subject<number>();
+
+    inflightEvents$.subscribe((value) => (inflightEventsCounter += value));
+
+    return upstream$.pipe(
+      // only take events for the configured channel
+      rx.filter((event) => event.channel === channel),
+
+      rx.switchMap((event) => {
+        if (inflightEventsCounter < this.getConfigFor(channel).inflightEventsThreshold) {
+          return rx.of(event);
+        }
+        this.logger.l(
+          `>> Dropping event ${event} (channel: ${channel}, inflightEventsCounter: ${inflightEventsCounter})`
+        );
+        this.senderUtils?.incrementCounter(TelemetryCounter.DOCS_DROPPED, 1, channel);
+
+        return rx.EMPTY;
+      }),
+
+      // update inflight events counter
+      rx.tap(() => {
+        inflightEvents$.next(1);
+      }),
+
+      // buffer events for a while
+      rx.bufferWhen<Event>(() => rx.interval(this.getConfigFor(channel).bufferTimeSpanMillis)),
+      // rx.bufferCount(3),
+
+      // exclude empty buffers
+      rx.filter((n: Event[]) => n.length > 0),
+
+      // serialize the payloads
+      // TODO(sebastian.zaffarano): configure serializer (e.g. protobuf)
+      rx.map((events) => events.map((e) => JSON.stringify(e.payload))),
+
+      // chunk by size
+      rx.map((values) =>
+        collections.chunkedBy(
+          values,
+          this.getConfigFor(channel).maxPayloadSizeBytes,
+          (payload) => payload.length
+        )
+      ),
+      rx.concatAll(),
+
+      // send events to the telemetry server
+      rx.concatMap((payloads: string[]) =>
+        retryOnError$(
+          this.getRetryConfig().retryCount,
+          this.getRetryConfig().retryDelayMillis,
+          async () => send(channel, payloads)
+        )
+      ),
+
+      // update inflight events counter
+      rx.tap((result: Result) => {
+        inflightEvents$.next(-result.events);
+      })
+    ) as rx.Observable<Result>;
+  }
+
+  private async sendEvents(channel: TelemetryChannel, events: string[]): Promise<Result> {
+    try {
+      const senderMetadata = await this.getSenderMetadata(channel);
+
+      const isTelemetryOptedIn = await senderMetadata.isTelemetryOptedIn();
+      if (!isTelemetryOptedIn) {
+        this.senderUtils?.incrementCounter(
+          TelemetryCounter.TELEMETRY_OPTED_OUT,
+          events.length,
+          channel
+        );
+
+        this.logger.l('Telemetry is not opted-in.');
+        return Promise.reject(newFailure('Telemetry is not opted-in', channel, events.length));
```

Comment:
any reason for `Promise.reject` over `throw newFailure...`?

Reply:
Not really.  To be honest, I don't remember why I did it :)
Updated to throw the exceptions instead of rejecting the promise.